x-pack/plugins/cloud_security_posture/public/pages/rules/rules_table.tsx

```diff
@@ -37,55 +50,166 @@ export const RulesTable = ({
+      background: row.metadata?.id === selectedRuleId ? euiTheme.colors.highlight : undefined,
     },
   });
 
+  const [selectAllRulesThisPage, setSelectAllRulesThisPage] = useState<boolean>(false);
+  const postRequestChangeRulesStatus = useChangeCspRuleStatus();
+
+  const isArraySubset = (
+    smallArr: CspBenchmarkRulesWithStatus[],
+    bigArr: CspBenchmarkRulesWithStatus[]
+  ) => {
+    let i: number = 0;
+    const newSmallArr = smallArr.map((e) => e.metadata);
+    const newBigArr = bigArr.map((e) => e.metadata);
+
+    while (i < newSmallArr.length) {
+      if (!newBigArr.includes(newSmallArr[i])) return false;
+      i++;
+    }
+    return true;
+  };
+
+  useEffect(() => {
+    if (selectedRules.length >= items.length) setSelectAllRulesThisPage(true);
+  }, [items.length, selectedRules.length]);
+
+  const columns = useMemo(
+    () =>
+      getColumns({
+        setSelectedRuleId,
+        refetchStatus,
+        postRequestChangeRulesStatus,
+        selectedRules,
+        setSelectedRules,
+        items,
+        setSelectAllRulesThisPage,
+        selectAllRulesThisPage,
+        isArraySubset,
+      }),
+    [
+      setSelectedRuleId,
+      refetchStatus,
+      postRequestChangeRulesStatus,
+      selectedRules,
+      setSelectedRules,
+      items,
+      setSelectAllRulesThisPage,
+      selectAllRulesThisPage,
+    ]
+  );
+
   return (
-    <EuiBasicTable
-      data-test-subj={TEST_SUBJECTS.CSP_RULES_TABLE}
-      loading={loading}
-      error={error}
-      items={items}
-      columns={columns}
-      pagination={euiPagination}
-      onChange={onTableChange}
-      itemId={(v) => v.metadata.id}
-      rowProps={rowProps}
-    />
+    <>
+      <EuiBasicTable
+        data-test-subj={TEST_SUBJECTS.CSP_RULES_TABLE}
+        loading={loading}
+        error={error}
+        items={items}
+        columns={columns}
+        pagination={euiPagination}
+        onChange={onTableChange}
+        itemId={(v) => v.metadata.id}
+        rowProps={rowProps}
+      />
+    </>
   );
 };
 
-type GetColumnProps = Pick<RulesTableProps, 'setSelectedRuleId'>;
-
 const getColumns = ({
   setSelectedRuleId,
-}: GetColumnProps): Array<EuiTableFieldDataColumnType<CspBenchmarkRule>> => [
+  refetchStatus,
+  postRequestChangeRulesStatus,
+  selectedRules,
+  setSelectedRules,
+  items,
+  setSelectAllRulesThisPage,
+  selectAllRulesThisPage,
+  isArraySubset,
+}: GetColumnProps & any): Array<EuiTableFieldDataColumnType<CspBenchmarkRulesWithStatus>> => [
+  {
+    field: 'action',
+    name: (
+      <EuiCheckbox
+        id={`cloud-security-fields-selector-item-all`}
+        checked={isArraySubset(items, selectedRules) && selectAllRulesThisPage}
+        onChange={(e) => {
+          const uniqueSelectedRules = uniqBy([...selectedRules, ...items], 'metadata.id');
+          const onChangeSelectAllThisPageFn = () => {
+            setSelectAllRulesThisPage(true);
```

Comment:
this sounds like something that doesn't requires a state, you can just derive if this page has selected rules by checking if the selected rules list contains items, lets remove this state